src/raftstore/store/fsm/peer.rs

```diff
@@ -1858,174 +1943,131 @@ impl<T: Transport, C: PdClient> Store<T, C> {
-                );
-                return;
-            }
-        };
-        peer.approximate_keys = Some(keys);
+    pub fn on_approximate_region_keys(&mut self, keys: u64) {
+        self.peer.approximate_keys = Some(keys);
+    }
+
+    pub fn on_compaction_declined_bytes(&mut self, ctx: &mut PollContext<T, C>, declined_bytes: u64) {
+        self.peer.compaction_declined_bytes += declined_bytes;
+        if self.peer.compaction_declined_bytes >= ctx.cfg.region_split_check_diff.0 {
+            UPDATE_REGION_SIZE_BY_COMPACTION_COUNTER.inc();
+        }
     }
 
     pub fn on_schedule_half_split_region(
         &mut self,
-        region_id: u64,
+        ctx: &mut PollContext<T, C>,
         region_epoch: &metapb::RegionEpoch,
         policy: CheckPolicy,
     ) {
-        let peer = match self.region_peers.get(&region_id) {
-            Some(peer) => peer,
-            None => {
-                error!("{:?}", Error::RegionNotFound(region_id));
-                return;
-            }
-        };
-
-        if !peer.is_leader() {
+        if !self.peer.is_leader() {
             // region on this store is no longer leader, skipped.
             warn!(
-                "[region {}] region on {} is not leader, skip.",
-                region_id,
-                self.store_id()
+                "{} is not leader, skip.",
+                self.peer.tag,
             );
             return;
         }
 
-        let region = peer.region();
+        let region = self.peer.region();
         if util::is_epoch_stale(region_epoch, region.get_region_epoch()) {
             warn!("[region {}] receive a stale halfsplit message", region_id);
             return;
         }
 
         let task = SplitCheckTask::new(region.clone(), false, policy);
-        if let Err(e) = self.split_check_worker.schedule(task) {
-            error!("{} failed to schedule split check: {}", self.tag, e);
+        if let Err(e) = ctx.split_check_scheduler.schedule(task) {
+            error!("{} failed to schedule split check: {}", self.peer.tag, e);
         }
     }
 
-    pub fn on_pd_heartbeat_tick(&mut self, event_loop: &mut EventLoop<Self>) {
-        for peer in self.region_peers.values_mut() {
-            peer.check_peers();
-        }
-        let mut leader_count = 0;
-        for peer in self.region_peers.values_mut() {
-            if peer.is_leader() {
-                leader_count += 1;
-                peer.heartbeat_pd(&self.pd_worker);
-            }
-        }
-        STORE_PD_HEARTBEAT_GAUGE_VEC
-            .with_label_values(&["leader"])
-            .set(leader_count);
-        STORE_PD_HEARTBEAT_GAUGE_VEC
-            .with_label_values(&["region"])
-            .set(self.region_peers.len() as i64);
+    pub fn on_pd_heartbeat_tick(&mut self, ctx: &mut PollContext<T, C>) {
+        self.peer.check_peers();
 
-        self.register_pd_heartbeat_tick(event_loop);
+        if !self.peer.is_leader() {
+            return;
```

Comment:
Also need to register_pd_heartbeat_tick when this peer is not leader. It's better to keep the previous style that register at first.